webapp/src/components/setup_ui/setup_ui.jsx

```diff
@@ -26,10 +25,7 @@ export default class SetupUI extends PureComponent {
     };
 
     componentDidMount() {
-        if (this.props.instanceInstalled && this.props.userConnected) {
-            this.registerHeaderButton();
-        }
-        setupUI();
+        this.props.setupUI();
```

Comment:
Won't this fail if the component is not updated after mounting? Maybe we need to make `headerButtonId` a global so it persists over mounts/dismount of our component? Then also check the state here?

Reply:
@crspeller I'm not sure whether it will or not, so I've changed the PR to track the header id in the plugin object, and pass it through like a value prop. This way we're not recreating a new one. It's not as bad as a global, and it's a bit more 'react-y'. Still feels weird to me.

But I found something concerning: the `setupUILater` function is being called everytime the component remounts. That's as intended. Problem is, the component is mounting a lot more often than I thought it was. And `setupUILater` is being called a lot more than I thought it would. But the icons aren't being duplicated. Why? `setupUILater` is stopping right at the call to `registry.registerReducer(reducers);`. So we aren't making new icons... by accident?

What do you think, should should I fix this to make it work more like this PR's solution? (Explicitly check if the icons are loaded with a plugin-level variable.)
Or should we just leave it because it seems to be working for whatever reason.

Reply:
I'm 1/5 on figuring out why it's being called more often than expected. Even though the result is what we want, having code that behaves a certain way without knowing why can result in tech debt and other bugs in the future.

I'm not sure what the implications are, for the function being called more than expected, but it makes me feel like we don't have control over the code. Does it make sense to store the initialization state in redux?

Reply:
@mickmister Made a ticket for it: https://mattermost.atlassian.net/browse/MM-17318

Reply:
@mickmister I think it doesn't make sense to store the state in redux. It's too heavy for something like this. To me, it makes sense to store state in the plugin's object. @crspeller ?

Reply:
0/5 on redux vs state. It doesn't really feel like shared state to me so I am fine with outside redux. 

> the component is mounting a lot more often than I thought it was

Welcome to react. :P